API Reference for Spend Permissions (#89)
coinbase/smart-wallet-docs: docs/pages/guides/spend-permissions/api-reference.mdx

| Field       | Type     | Description                                                                 |
|-------------|----------|-----------------------------------------------------------------------------|
| `account`   | `address`| Smart account this spend permission is valid for.                           |
| `spender`   | `address`| Entity that can spend `account`'s tokens.                                   |
| `token`     | `address`| Token address (ERC-7528 native token address or ERC-20 contract).           |
| `allowance` | `uint160`| Maximum allowed value to spend within each `period`.                        |
| `period`    | `uint48` | Time duration for resetting used `allowance` on a recurring basis (seconds).|
| `start`     | `uint48` | Timestamp this spend permission is valid after (unix seconds).              |
| `end`       | `uint48` | Timestamp this spend permission is valid until (unix seconds).              |
| `salt`      | `uint256`| An arbitrary salt to differentiate unique spend permissions with otherwise identical data. |
| `extraData` | `bytes`  | Arbitrary data to include in the signature.                                 |

---

> #### `SpendPermissionBatch`

Reviewer: would skip any documentation for batching behavior until Keys web is ready to handle it

| `spender`   | `address`| Entity that can spend user funds.                                           |
| `token`     | `address`| Token address (ERC-7528 ether address or ERC-20 contract).                  |
| `allowance` | `uint160`| Maximum allowed value to spend within a recurring period.                   |
| `salt`      | `uint256`| An arbitrary salt to differentiate unique spend permissions with otherwise identical data. |
| `extraData` | `bytes`  | Arbitrary data to include in the signature.                                 |

---

#### `PeriodSpend`

| Field       | Type     | Description                                                                 |
|-------------|----------|-----------------------------------------------------------------------------|
| `start`     | `uint48` | Start time of the period (unix seconds).                                    |
| `end`       | `uint48` | End time of the period (unix seconds).                                      |
| `spend`     | `uint160`| Accumulated spend amount for period.                                        |

---

### Contract functions
#### `approve`

> Approve a spend permission via a direct call from the account. Only callable by the account owner specified in the spend permission.

Reviewer: ```suggestion
Approve a spend permission via a direct call from the `account`. Only callable by the `account` specified in the spend permission.
```

```solidity

> function approve(SpendPermission calldata spendPermission) external requireSender(spendPermission.account);

Reviewer: Would remove the `requireSender` modifier for cleanliness and your comment explains the behavior already which is good

```

---

#### `approveWithSignature`

> Approve a spend permission via a signature from the account owner. Compatible with [ERC-6492](https://eips.ethereum.org/EIPS/eip-6492) signatures.

Reviewer: ```suggestion
Approve a spend permission via a signature from the `account`. Compatible with [ERC-6492](https://eips.ethereum.org/EIPS/eip-6492) signatures for automatic account creation if needed.
```

```solidity
function approveWithSignature(SpendPermission calldata spendPermission, bytes calldata signature) external;
```

---

> #### `approveBatchWithSignature`

Reviewer: prune any batching for now

Approve a batch of spend permissions via one signature from the account owner. Batched permissions share an `account`, `period`,
`start` and `end`. Details unique to each permission in the batch are specified by an array of `PermissionDetails` structs.

```solidity
function approveBatchWithSignature(SpendPermissionBatch memory spendPermissionBatch, bytes calldata signature) external;
```

---

#### `spend`

> Spend tokens using a spend permission, transferring them from the account to the spender. Only callable by the spender specified in the permission.

Reviewer: ```suggestion
Spend tokens using a spend permission, transferring them from the `account` to the `spender`. Only callable by the `spender` specified in the permission.
```

```solidity
function spend(SpendPermission memory spendPermission, uint160 value) external requireSender(spendPermission.spender);
```

---

#### `revoke`

> Revoke a spend permission, permanently disabling its use. Only callable by the account owner specified in the spend permission.

Reviewer: ```suggestion
Revoke a spend permission, permanently disabling its use. Only callable by the `account` specified in the spend permission.
```

```solidity
function revoke(SpendPermission calldata spendPermission) external requireSender(spendPermission.account);
```

---

#### `getHash`

Generate a hash of a `SpendPermission` struct for signing, in accordance with [EIP-712](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md).

```solidity
function getHash(SpendPermission memory spendPermission) public view returns (bytes32);
```

---

> #### `getBatchHash`

Reviewer: prune batch for now

Generate a hash of a `SpendPermissionBatch` struct for signing, in accordance with [EIP-712](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md). 
Reverts if the batch is empty.

```solidity
function getBatchHash(SpendPermissionBatch memory spendPermissionBatch) public view returns (bytes32);
```

---

#### `isApproved`

Check if a spend permission is approved and not revoked.

```solidity
function isApproved(SpendPermission memory spendPermission) public view returns (bool);
```

---

#### `getCurrentPeriod`

> Retrieve the start, end, and accumulated spend for the current period of a spend permission.

Reviewer: ```suggestion
Retrieve the `start`, `end`, and accumulated `spend` for the current period of a spend permission.
```

```solidity
function getCurrentPeriod(SpendPermission memory spendPermission) public view returns (PeriodSpend memory);
```
## Coinbase Smart Wallet clients

### Using Viem
[Viem](https://viem.sh/) is a TypeScript interface for Ethereum that provides low-level stateless primitives for interacting with Ethereum.

- **[Getting Started with Viem](https://viem.sh/docs/getting-started.html)** 

- **[Getting Started with Account Abstraction](https://viem.sh/account-abstraction#getting-started-with-account-abstraction)** 
   Viem provides support for account abstraction clients, including specific support for Coinbase Smart Wallet and paymasters.

### Using OnchainKit
[OnchainKit](https://onchainkit.xyz/) is a collection of React components and TypeScript utilities that help developers quickly build
onchain applications.

### Using Wagmi
[Wagmi](https://wagmi.sh/) is a collection of React Hooks that facilitate development of blockchain frontends.

-  **[Getting Started with Wagmi](https://wagmi.sh/react/getting-started)** 
- **[coinbaseWallet](https://wagmi.sh/react/api/connectors/coinbaseWallet)** 

> ## Coinbase Wallet API

Reviewer: would recommend splitting API and contracts into separate interface reference pages to keep each page digestible and scoped

> ### `wallet_fetchPermissions`

Reviewer: Need to declare the endpoint this is accessible at and also emphasize that this is JSON-RPC style

#### Schema
```typescript
type FetchPermissionsRequest = {
  account: Address;
  chainId: string; // 0x...
  spender: Address;

>   pageOptions: PageOptions; // optional, defaults to page size 50 and empty cursor

Reviewer: ```suggestion
  pageOptions?: PageOptions; // optional, defaults to page size 50 and empty cursor
```

```
## Coinbase Smart Wallet clients

### Using Viem
[Viem](https://viem.sh/) is a TypeScript interface for Ethereum that provides low-level stateless primitives for interacting with Ethereum.

- **[Getting Started with Viem](https://viem.sh/docs/getting-started.html)** 

- **[Getting Started with Account Abstraction](https://viem.sh/account-abstraction#getting-started-with-account-abstraction)** 
   Viem provides support for account abstraction clients, including specific support for Coinbase Smart Wallet and paymasters.

### Using OnchainKit
[OnchainKit](https://onchainkit.xyz/) is a collection of React components and TypeScript utilities that help developers quickly build
onchain applications.

### Using Wagmi
[Wagmi](https://wagmi.sh/) is a collection of React Hooks that facilitate development of blockchain frontends.

-  **[Getting Started with Wagmi](https://wagmi.sh/react/getting-started)** 
- **[coinbaseWallet](https://wagmi.sh/react/api/connectors/coinbaseWallet)** 

## Coinbase Wallet API

### `wallet_fetchPermissions`

#### Schema
```typescript
type FetchPermissionsRequest = {

>   account: Address;
>   chainId: string; // 0x...
>   spender: Address;

Reviewer: `Address` is inconsistent with rest of style here. Can we just make them all string with a comment for the actual type and expected format.
```suggestion
  chainId: string; // uint256, hex formatted
  account: string; // address
  spender: string; // address
```

  pageOptions: PageOptions; // optional, defaults to page size 50 and empty cursor
}

> type FetchPermissionsResponse = {

Reviewer: can we use naming convention of "result" to match typical JSON-RPC style?
```suggestion
type FetchPermissionsResult = {
```

  permissions: FetchPermissionsResponseItem[];
  pageDescription: PageDescription;
}

type PageOptions = {
  pageSize number; // number of items requested, defaults to 50
  cursor string; // identifier for where the page should start
}

type PageDescription = {
  pageSize number; // number of items returned
  nextCursor string; // identifier for where the next page should start

> }

Reviewer: I get why these are nested into their own objects, but it just adds a bit more clutter for someone trying to see the whole shape of the API to have separate types. Can we remove these types and have the inner object in the other types, e.g.
```typescript
type FetchPermissionsResult = {
  permissions: FetchPermissionsResultItem[];
  pageDescription: {
    pageSize: number;
    nextCursor: string;
  }
}

type FetchPermissionsResponseItem = {
  createdAt: number; // UTC timestamp for when the permission was granted
  permissionHash: string;
  signature: string;
  permission: {
    account: string;
    spender: string;

>     token: string;

Reviewer: add comments annotating strings with expected shape e.g. address/bytes32/hex

    allowance: string; // base 10 numeric string
    period: number;
    start: number;
    end: number;
    salt: string; // base 10 numeric string
    extraData: string // 0x...
  };
}
```

#### Example request

```bash

> curl --location 'https://wallet-chain-proxy-dev.cbhq.net?targetName=base-sepolia' \

Reviewer: Gut says we shouldn't be exposing this endpoint, can you confirm with Sam?

Reviewer 2: This is our internal dev endpoint, for now let's keep this out of the docs since devs won't be able to hit this externally until we launch to prod. For reference, this will eventually be `https://chain-proxy.wallet.coinbase.com/?targetName=base-sepolia` though.

Author: great, just pulling out the entire example request/response for now

--header 'Content-Type: application/json' \
--data '{
    "jsonrpc":"2.0",
    "method":"wallet_fetchPermissions",

>     "params":["0x764e97e81a0b1138c70525499037e27e442bd3b2","0x14A34","0xd4e17478581878a967aa22d45a5158a9fe96aa08"],

Reviewer: this sample request is of different type than what was laid out above

Reviewer 2: I'm not familiar enough with JSON RPCs and how we typically document them, but this is how we are expecting the payload in the backend.
